apps/golden-sample-app/src/app/user-context/user-context.guard.ts

```diff
@@ -0,0 +1,71 @@
+import { Injectable } from '@angular/core';
+import {
+  ActivatedRouteSnapshot,
+  CanActivate,
+  CanActivateChild,
+  Router,
+  RouterStateSnapshot,
+  UrlTree,
+} from '@angular/router';
+import { ServiceAgreementHttpService } from '@backbase/data-ang/accesscontrol';
+import { CookieService } from 'ngx-cookie-service';
+import { Observable, of } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class UserContextGuard implements CanActivate, CanActivateChild {
+  private targetUrl: string | undefined;
+  private cookieValid = false;
+
+  constructor(
+    private readonly cookieService: CookieService,
+    private readonly router: Router,
+    private readonly serviceAgreementService: ServiceAgreementHttpService,
+  ) { }
+  canActivateChild(
+    childRoute: ActivatedRouteSnapshot,
+    state: RouterStateSnapshot,
+  ): boolean | UrlTree | Observable<boolean | UrlTree> | Promise<boolean | UrlTree> {
+    return this.canActivate(childRoute, state);
+  }
+
+  canActivate(
+    _next: ActivatedRouteSnapshot,
+    state: RouterStateSnapshot,
+  ): Observable<boolean | UrlTree> | Promise<boolean | UrlTree> | boolean | UrlTree {
+    this.targetUrl = state.url;
+    const cookieIsSet = this.cookieService.check('USER_CONTEXT');
+    console.log('cookieIsSet', cookieIsSet);
+    if (!cookieIsSet) {
+      return this.getSelectContextUrlTree();
+    }
+
+    return (
+      this.cookieValid ||
+      this.validateUserContextCookie().pipe(map(isValid => isValid || this.getSelectContextUrlTree()))
+    );
+  }
```

Comment:
Maybe explain the logic in comment, it is difficult to understand why and what are you doing here.

Reply:
I've rewritten this guard a bit. @igord could you please review it again?